File: SLAM/TrimForm.vb
Imports NAudio
Imports NAudio.Wave
Imports System.Threading
Imports System.IO

Public Class TrimForm
    Public WavFile As String
    Public startpos As Integer
    Public endpos As Integer

    Private Sub TrimForm_Load(sender As Object, e As EventArgs) Handles MyBase.Load
        If Not String.IsNullOrEmpty(WavFile) Then
            Using reader As New Wave.WaveFileReader(WavFile)
                AdvWaveViewer1.WaveStream = New Wave.WaveFileReader(WavFile)
                'reader.WaveFormat.SampleRate
            End Using

            NumericRightS.Maximum = Decimal.MaxValue
            NumericRight.Maximum = AdvWaveViewer1.MaxSamples
            NumericRight.Increment = AdvWaveViewer1.SamplesPerPixel

            NumericLeftS.Maximum = Decimal.MaxValue
            NumericLeft.Maximum = Decimal.MaxValue
            NumericLeft.Increment = AdvWaveViewer1.SamplesPerPixel

            If startpos = endpos And endpos = 0 Then
                NumericRight.Value = AdvWaveViewer1.MaxSamples
            Else
                AdvWaveViewer1.rightpos = endpos
                AdvWaveViewer1.leftpos = startpos
                NumericRight.Value = AdvWaveViewer1.rightSample
                NumericLeft.Value = AdvWaveViewer1.leftSample
            End If

        End If
    End Sub

    Private Sub TrimForm_FormClosing(sender As Object, e As FormClosingEventArgs) Handles Me.FormClosing
        If BackgroundPlayer.IsBusy Then
            BackgroundPlayer.CancelAsync()
        End If
    End Sub

    Private Sub AdvWaveViewer1_MouseUp(sender As Object, e As MouseEventArgs) Handles AdvWaveViewer1.MouseUp
        NumericRight.Value = AdvWaveViewer1.rightSample
        NumericLeft.Value = AdvWaveViewer1.leftSample
    End Sub

    Private Sub NumericLeft_ValueChanged(sender As Object, e As EventArgs) Handles NumericLeft.ValueChanged
        If NumericLeft.Value >= AdvWaveViewer1.rightSample Then
            NumericLeft.Value = NumericRight.Value - 1
        End If
        AdvWaveViewer1.leftSample = NumericLeft.Value
        NumericLeftS.Value = NumericLeft.Value / AdvWaveViewer1.SampleRate
    End Sub

    Private Sub NumericRight_ValueChanged(sender As Object, e As EventArgs) Handles NumericRight.ValueChanged
        If NumericRight.Value <= AdvWaveViewer1.leftSample Then
            NumericRight.Value = NumericLeft.Value + 1
        End If
        AdvWaveViewer1.rightSample = NumericRight.Value
        NumericRightS.Value = NumericRight.Value / AdvWaveViewer1.SampleRate
    End Sub

    Private Sub DoneButton_Click(sender As Object, e As EventArgs) Handles DoneButton.Click
        startpos = AdvWaveViewer1.leftpos

        If AdvWaveViewer1.rightSample = AdvWaveViewer1.MaxSamples And AdvWaveViewer1.leftpos = 0 Then
            endpos = 0
        Else
            endpos = AdvWaveViewer1.rightpos
        End If


        DialogResult = Windows.Forms.DialogResult.OK
    End Sub

    Private Sub TrimForm_Resize(sender As Object, e As EventArgs) Handles Me.Resize
        NumericLeft.Increment = AdvWaveViewer1.SamplesPerPixel
        NumericRight.Increment = AdvWaveViewer1.SamplesPerPixel
    End Sub

    Private Sub ResetButton_Click(sender As Object, e As EventArgs) Handles ResetButton.Click
        NumericLeft.Value = 0
        NumericRight.Value = AdvWaveViewer1.MaxSamples
    End Sub

    Private Sub NumericLeftS_ValueChanged(sender As Object, e As EventArgs) Handles NumericLeftS.ValueChanged
        NumericLeft.Value = NumericLeftS.Value * AdvWaveViewer1.SampleRate
    End Sub

    Private Sub NumericRightS_ValueChanged(sender As Object, e As EventArgs) Handles NumericRightS.ValueChanged
        If (NumericRightS.Value * AdvWaveViewer1.SampleRate) <= NumericRight.Maximum Then
            NumericRight.Value = NumericRightS.Value * AdvWaveViewer1.SampleRate
        Else
            NumericRight.Value = NumericRight.Maximum
            NumericRightS.Value = NumericRight.Value / AdvWaveViewer1.SampleRate
        End If

    End Sub

    Private Sub PlayButton_Click(sender As Object, e As EventArgs) Handles PlayButton.Click
        If BackgroundPlayer.IsBusy = False Then
            BackgroundPlayer.RunWorkerAsync(New Object(2) {AdvWaveViewer1.WaveStream, AdvWaveViewer1.leftpos, AdvWaveViewer1.rightpos})
            PlayButton.Text = "Stop"
            DisableInterface()
            PlayButton.Enabled = True
        Else
            BackgroundPlayer.CancelAsync()
            PlayButton.Text = "Play"
        End If
    End Sub

    Private Sub BackgroundPlayer_DoWork(sender As Object, e As System.ComponentModel.DoWorkEventArgs) Handles BackgroundPlayer.DoWork
        Dim WaveFloat As WaveStream = e.Argument(0)
        Dim LeftPos As Integer = e.Argument(1)
        Dim RightPos As Integer = e.Argument(2)

        Dim bytes((RightPos - LeftPos)) As Byte

        WaveFloat.Position = LeftPos
        WaveFloat.Read(bytes, 0, (RightPos - LeftPos))

        WaveFloat = New RawSourceWaveStream(New MemoryStream(bytes), WaveFloat.WaveFormat)
        'WaveFloat.PadWithZeroes = False

        Using output = New WaveOutEvent()
            output.Init(WaveFloat)
            output.Play()
            While output.PlaybackState = PlaybackState.Playing And Not BackgroundPlayer.CancellationPending
                Thread.Sleep(45)
                BackgroundPlayer.ReportProgress(output.GetPosition() / (WaveFloat.WaveFormat.BitsPerSample / 8))
            End While
        End Using
    End Sub

    Private Sub BackgroundPlayer_ProgressChanged(sender As Object, e As System.ComponentModel.ProgressChangedEventArgs) Handles BackgroundPlayer.ProgressChanged
        AdvWaveViewer1.marker = e.ProgressPercentage
    End Sub

    Private Sub BackgroundPlayer_RunWorkerCompleted(sender As Object, e As System.ComponentModel.RunWorkerCompletedEventArgs) Handles BackgroundPlayer.RunWorkerCompleted
        PlayButton.Text = "Play"
        AdvWaveViewer1.marker = 0
        EnableInterface()
    End Sub

    Private Sub EnableInterface()
        For Each Control In Me.Controls
            Control.Enabled = True
        Next
    End Sub

    Private Sub DisableInterface()
        For Each Control In Me.Controls
            Control.Enabled = False
        Next
    End Sub
End Class

File: SLAM/TrimForm.cs
using System;
using System.IO;
using System.Threading;
using System.Windows.Forms;
using Microsoft.VisualBasic.CompilerServices;
using NAudio.Wave;

namespace SLAM
{
    public partial class TrimForm
    {
        public TrimForm()
        {
            InitializeComponent();
            _NumericRight.Name = "NumericRight";
            _DoneButton.Name = "DoneButton";
            _NumericLeft.Name = "NumericLeft";
            _ResetButton.Name = "ResetButton";
            _NumericLeftS.Name = "NumericLeftS";
            _NumericRightS.Name = "NumericRightS";
            _PlayButton.Name = "PlayButton";
            _AdvWaveViewer1.Name = "AdvWaveViewer1";
        }

        public string WavFile;
        public int startpos;
        public int endpos;

        private void TrimForm_Load(object sender, EventArgs e)
        {
            if (!string.IsNullOrEmpty(WavFile))
            {
                using (var reader = new WaveFileReader(WavFile))
                {
                    AdvWaveViewer1.WaveStream = new WaveFileReader(WavFile);
                    // reader.WaveFormat.SampleRate
                }

                NumericRightS.Maximum = decimal.MaxValue;
                NumericRight.Maximum = AdvWaveViewer1.MaxSamples;
                NumericRight.Increment = AdvWaveViewer1.SamplesPerPixel;
                NumericLeftS.Maximum = decimal.MaxValue;
                NumericLeft.Maximum = decimal.MaxValue;
                NumericLeft.Increment = AdvWaveViewer1.SamplesPerPixel;
                if (startpos == endpos & endpos == 0)
                {
                    NumericRight.Value = AdvWaveViewer1.MaxSamples;
                }
                else
                {
                    AdvWaveViewer1.rightpos = endpos;
                    AdvWaveViewer1.leftpos = startpos;
                    NumericRight.Value = AdvWaveViewer1.rightSample;
                    NumericLeft.Value = AdvWaveViewer1.leftSample;
                }
            }
        }

        private void TrimForm_FormClosing(object sender, FormClosingEventArgs e)
        {
            if (BackgroundPlayer.IsBusy)
            {
                BackgroundPlayer.CancelAsync();
            }
        }

        private void AdvWaveViewer1_MouseUp(object sender, MouseEventArgs e)
        {
            NumericRight.Value = AdvWaveViewer1.rightSample;
            NumericLeft.Value = AdvWaveViewer1.leftSample;
        }

        private void NumericLeft_ValueChanged(object sender, EventArgs e)
        {
            if (NumericLeft.Value >= AdvWaveViewer1.rightSample)
            {
                NumericLeft.Value = NumericRight.Value - 1m;
            }

            AdvWaveViewer1.leftSample = (int)NumericLeft.Value;
            NumericLeftS.Value = NumericLeft.Value / AdvWaveViewer1.SampleRate;
        }

        private void NumericRight_ValueChanged(object sender, EventArgs e)
        {
            if (NumericRight.Value <= AdvWaveViewer1.leftSample)
            {
                NumericRight.Value = NumericLeft.Value + 1m;
            }

            AdvWaveViewer1.rightSample = (int)NumericRight.Value;
            NumericRightS.Value = NumericRight.Value / AdvWaveViewer1.SampleRate;
        }

        private void DoneButton_Click(object sender, EventArgs e)
        {
            startpos = AdvWaveViewer1.leftpos;
            if (AdvWaveViewer1.rightSample == AdvWaveViewer1.MaxSamples & AdvWaveViewer1.leftpos == 0)
            {
                endpos = 0;
            }
            else
            {
                endpos = AdvWaveViewer1.rightpos;
            }

            DialogResult = DialogResult.OK;
        }

        private void TrimForm_Resize(object sender, EventArgs e)
        {
            NumericLeft.Increment = AdvWaveViewer1.SamplesPerPixel;
            NumericRight.Increment = AdvWaveViewer1.SamplesPerPixel;
        }

        private void ResetButton_Click(object sender, EventArgs e)
        {
            NumericLeft.Value = 0m;
            NumericRight.Value = AdvWaveViewer1.MaxSamples;
        }

        private void NumericLeftS_ValueChanged(object sender, EventArgs e)
        {
            NumericLeft.Value = NumericLeftS.Value * AdvWaveViewer1.SampleRate;
        }

        private void NumericRightS_ValueChanged(object sender, EventArgs e)
        {
            if (NumericRightS.Value * AdvWaveViewer1.SampleRate <= NumericRight.Maximum)
            {
                NumericRight.Value = NumericRightS.Value * AdvWaveViewer1.SampleRate;
            }
            else
            {
                NumericRight.Value = NumericRight.Maximum;
                NumericRightS.Value = NumericRight.Value / AdvWaveViewer1.SampleRate;
            }
        }

        private void PlayButton_Click(object sender, EventArgs e)
        {
            if (BackgroundPlayer.IsBusy == false)
            {
                BackgroundPlayer.RunWorkerAsync(new object[3] { AdvWaveViewer1.WaveStream, AdvWaveViewer1.leftpos, AdvWaveViewer1.rightpos });
                PlayButton.Text = "Stop";
                DisableInterface();
                PlayButton.Enabled = true;
            }
            else
            {
                BackgroundPlayer.CancelAsync();
                PlayButton.Text = "Play";
            }
        }

        private void BackgroundPlayer_DoWork(object sender, System.ComponentModel.DoWorkEventArgs e)
        {
            object[] arguments = e.Argument as object[];

            WaveStream WaveFloat = (WaveStream)arguments[0];
            int LeftPos = Conversions.ToInteger(arguments[1]);
            int RightPos = Conversions.ToInteger(arguments[2]);
            var bytes = new byte[RightPos - LeftPos + 1];
            WaveFloat.Position = LeftPos;
            WaveFloat.Read(bytes, 0, RightPos - LeftPos);
            WaveFloat = new RawSourceWaveStream(new MemoryStream(bytes), WaveFloat.WaveFormat);
            // WaveFloat.PadWithZeroes = False

            using (var output = new WaveOutEvent())
            {
                output.Init(WaveFloat);
                output.Play();
                while (output.PlaybackState == PlaybackState.Playing & !BackgroundPlayer.CancellationPending)
                {
                    Thread.Sleep(45);
                    BackgroundPlayer.ReportProgress((int)(output.GetPosition() / (WaveFloat.WaveFormat.BitsPerSample / 8d)));
                }
            }
        }

        private void BackgroundPlayer_ProgressChanged(object sender, System.ComponentModel.ProgressChangedEventArgs e)
        {
            AdvWaveViewer1.marker = e.ProgressPercentage;
        }

        private void BackgroundPlayer_RunWorkerCompleted(object sender, System.ComponentModel.RunWorkerCompletedEventArgs e)
        {
            PlayButton.Text = "Play";
            AdvWaveViewer1.marker = 0L;
            EnableInterface();
        }

        private void EnableInterface()
        {
            foreach (Control Control in Controls)
                Control.Enabled = true;
        }

        private void DisableInterface()
        {
            foreach (Control Control in Controls)
                Control.Enabled = false;
        }
    }
}